### PR TITLE
Add `nostd::make_unique`

### LINKED ISSUE #3096: Can not compile when using `WITH_STL=CXX11` with legacy toolchain
**Describe your environment** 

OS: Linux
Compiler: GCC 4.8.5
otel-cpp version: 1.17.0

**Steps to reproduce**
Building otel-cpp 1.17.0

**What is the expected behavior?**
Building success.

**What is the actual behavior?**

```bash
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.17.0/exporters/memory/src/in_memory_metric_exporter_factory.cc: In member function ‘virtual opentelemetry::v1::sdk::common::ExportResult opentelemetry::v1::exporter::memory::{anonymous}::InMemoryMetricExporter::Export(const opentelemetry::v1::sdk::metrics::ResourceMetrics&)’:
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.17.0/exporters/memory/src/in_memory_metric_exporter_factory.cc:52:16: error: ‘make_unique’ is not a member of ‘std’
     data_->Add(std::make_unique<ResourceMetrics>(data));
                ^
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.17.0/exporters/memory/src/in_memory_metric_exporter_factory.cc:52:48: error: expected primary-expression before ‘>’ token
     data_->Add(std::make_unique<ResourceMetrics>(data));

// ...
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.17.0/exporters/memory/src/in_memory_metric_exporter_factory.cc:82:26: error: parameter declared ‘auto’
   return Create(data, [](auto) { return AggregationTemporality::kCumulative; });


// ...
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.17.0/exporters/memory/src/in_memory_metric_data.cc: In member function ‘virtual void opentelemetry::v1::exporter::memory::SimpleAggregateInMemoryMetricData::Add(std::unique_ptr<opentelemetry::v1::sdk::metrics::ResourceMetrics>)’:
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.17.0/exporters/memory/src/in_memory_metric_data.cc:34:30: error: converting to ‘std::map<std::tuple<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::map<opentelemetry::v1::sdk::common::OrderedAttributeMap, absl::lts_2020_02_25::variant<opentelemetry::v1::sdk::metrics::SumPointData, opentelemetry::v1::sdk::metrics::HistogramPointData, opentelemetry::v1::sdk::metrics::LastValuePointData, opentelemetry::v1::sdk::metrics::DropPointData> > >::key_type {aka std::tuple<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::basic_string<char, std::char_traits<char>, std::allocator<char> > >}’ from initializer list would use explicit constructor ‘constexpr std::tuple<_T1, _T2>::tuple(const _T1&, const _T2&) [with _T1 = std::basic_string<char>; _T2 = std::basic_string<char>]’
         data_[{scope, metric}].insert({pda.attributes, pda.point_data});
                              ^
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.17.0/exporters/memory/src/in_memory_metric_data.cc: In member function ‘const AttributeToPoint& opentelemetry::v1::exporter::memory::SimpleAggregateInMemoryMetricData::Get(const string&, const string&)’:
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.17.0/exporters/memory/src/in_memory_metric_data.cc:44:31: error: converting to ‘std::map<std::tuple<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::map<opentelemetry::v1::sdk::common::OrderedAttributeMap, absl::lts_2020_02_25::variant<opentelemetry::v1::sdk::metrics::SumPointData, opentelemetry::v1::sdk::metrics::HistogramPointData, opentelemetry::v1::sdk::metrics::LastValuePointData, opentelemetry::v1::sdk::metrics::DropPointData> > >::key_type {aka std::tuple<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::basic_string<char, std::char_traits<char>, std::allocator<char> > >}’ from initializer list would use explicit constructor ‘constexpr std::tuple<_T1, _T2>::tuple(const _T1&, const _T2&) [with _T1 = std::basic_string<char>; _T2 = std::basic_string<char>]’
   return data_[{scope, metric}];                    ^
```

### PR DESCRIPTION
**Pull request by owent**
Fixes #3096

## Changes

+ Add `nostd::make_unique`

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

File: exporters/memory/src/in_memory_metric_exporter_factory.cc [hunk 1]
```diff
@@ -49,7 +50,7 @@ class InMemoryMetricExporter final : public sdk::metrics::PushMetricExporter
       OTEL_INTERNAL_LOG_ERROR("[In Memory Metric Exporter] Exporting failed, exporter is shutdown");
       return ExportResult::kFailure;
     }
-    data_->Add(std::make_unique<ResourceMetrics>(data));
+    data_->Add(opentelemetry::nostd::make_unique<ResourceMetrics>(data));
```
[the remark this answers is not in the document]
Comment (owent) [reply exporters/memory/src/in_memory_metric_exporter_factory.cc:53]: Thanks, I modified the parameter type of this API but kept `std::unique_ptr` in the exported APIs for ABI compatibility.

**Comment by lalitb** [comment]
I'm a bit puzzled—why are we reintroducing support for building with GCC 4.8 and C++11 when we've already moved on to supporting C++14 and above? Or may be I am missing something :)

**Comment by marcalff** [comment]
Instead of reintroducing nostd helpers for code in the sdk already using `std::unique_ptr`, I think it will be better to settle on a subset of C++ 14.

Let's call it`C++13.9` ... almost as `C++14`, but we restrict ourselves to not use `std::make_unique`, so that it can also build in `C++11`.

There are 4 lines to fix in the entire code base, so that:

```
data_->Add(std::make_unique<ResourceMetrics>(data));
```

needs to change to

```
data_->Add(std::unique_ptr<ResourceMetrics>(new ResourceMetrics(data)));
```

and then very old toolchains will still work.

While C++11 is no longer officially supported, we should not break it if it is easy (4 lines) to avoid.

@owent would this work for you ?

**Comment by marcalff** [comment]
Previous discussion on C++11 / C++14:

https://github.com/open-telemetry/opentelemetry-cpp/pull/2522#issuecomment-1950475535

**Comment by ThomsonTan** [comment]
Yes, prefer not introducing `nostd::make_unique` as the C++ 14 is already the minimum requirement. And we also don't want to declare that it is compatible with `std::make_unique` in C++ 14 and above, like `nostd::make_unique(nostd::size_t size)` is not supported. So could outline the current code like @marcalff mentioned, or give it some other utility name if it is really necessary. This way makes it more flexible than polyfill a standard library method.

**Comment by owent** [comment]
> Instead of reintroducing nostd helpers for code in the sdk already using `std::unique_ptr`, I think it will be better to settle on a subset of C++ 14.
> 
> Let's call it`C++13.9` ... almost as `C++14`, but we restrict ourselves to not use `std::make_unique`, so that it can also build in `C++11`.
> 
> There are 4 lines to fix in the entire code base, so that:
> 
> ```
> data_->Add(std::make_unique<ResourceMetrics>(data));
> ```
> 
> needs to change to
> 
> ```
> data_->Add(std::unique_ptr<ResourceMetrics>(new ResourceMetrics(data)));
> ```
> 
> and then very old toolchains will still work.
> 
> While C++11 is no longer officially supported, we should not break it if it is easy (4 lines) to avoid.
> 
> @owent would this work for you ?

Yes, it works. But I wonder if it's necessary to keep `nostd::unique_ptr<T>` if we always depend `std::unique_ptr` ?

**Comment by marcalff** [comment]
> Yes, it works. But I wonder if it's necessary to keep `nostd::unique_ptr<T>` if we always depend `std::unique_ptr` ?
> 
> I raised #3098 to replace `std::make_unique` only. And this PR can be closed if we use #3098 .

In my understanding, `nostd::unique_ptr<T>` is now always implemented using `std::unique_ptr` internally (because we use C++ >= 11), but the prototype in the ABI only refers to `nostd::unique_ptr<T>`.

If we change the code in the API to replace nostd with std, the methods signatures will change, breaking the ABI.

**Pull request closed without merging**